akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -0,0 +1,254 @@
+
+    override def apply(t: Try[O]): Unit = {
+      setOutgoing(t)
+      cb.invoke(this)
+    }
+
+    override def toString: String = s"Holder($incoming, $outgoing, $partition)"
+  }
+
+  val NotYetThere = MapAsync.NotYetThere
+}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[akka] final case class MapAsyncPartitioned[In, Out, Partition](
+    parallelism: Int,
+    perPartition: Int,
+    partitioner: In => Partition,
+    f: (In, Partition) => Future[Out])
+    extends GraphStage[FlowShape[In, Out]] {
+  import MapAsyncPartitioned._
+
+  // duplicating the checks of the apply method since the constructor is public in the bytecode
+  if (parallelism < 1) throw new IllegalArgumentException("parallelism must be at least 1")
+  if (perPartition < 1) throw new IllegalArgumentException("perPartition must be at least 1")
+  if (perPartition >= parallelism)
+    throw new IllegalArgumentException("perPartition must be less than parallelism: consider mapAsync instead")
+
+  private val in = Inlet[In]("MapAsyncPartitioned.in")
+  private val out = Outlet[Out]("MapAsyncPartitioned.out")
+
+  override def initialAttributes = DefaultAttributes.mapAsync and SourceLocation.forLambda(f)
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+      var buffer: PartitionedBuffer[Partition, Holder[Partition, In, Out]] = _
+
+      private val futureCB = getAsyncCallback[Holder[Partition, In, Out]] { holder =>
+        holder.outgoing match {
+          case Failure(ex) if holder.supervisionDirectiveFor(decider, ex) == Supervision.Stop => failStage(ex)
+          case _ =>
+            dropCompletedThenPushIfPossible(holder.partition)
+        }
+      }
+
+      override def preStart(): Unit = {
+        buffer = new PartitionedBuffer[Partition, Holder[Partition, In, Out]](
+          linearBuffer = BufferImpl(parallelism, inheritedAttributes),
+          partitioner = _.partition,
+          makePartitionBuffer = { (p, overflowCapacity) =>
+            import akka.stream.impl.{ BoundedBuffer, ChainedBuffer }
+            val tailCapacity = (overflowCapacity - perPartition).max(1)
+
+            new ChainedBuffer[Holder[Partition, In, Out]](
+              BufferImpl(perPartition, inheritedAttributes),
+              new BoundedBuffer.DynamicQueue(tailCapacity), {
+                holder =>
+                  // this will execute when the holder moves from the tail buffer to the head buffer (viz. ready to be
+                  //  scheduled).  This might be on enqueueing into the partitioned buffer, or it might be when
+                  //  the head of this partition is dropped after being completed.
+                  try {
+                    val future = f(holder.incoming, p)
+                    holder.clearIncoming()
```

Comment:
```suggestion
                    // optimization, clear the in-element reference once we have used it so it can be gc:d
                    holder.clearIncoming()
```